File: src/main/kotlin/com/lcaohoanq/sp/domains/notifications/NotificationPort.java
package com.lcaohoanq.sp.domains.notifications;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.lcaohoanq.sp.domains.notifications.NotificationEntity.NotificationColor;
import com.lcaohoanq.sp.domains.notifications.NotificationEntity.NotificationIcon;
import com.lcaohoanq.sp.domains.notifications.NotificationEntity.NotificationType;
import java.time.LocalDateTime;

public interface NotificationPort {

    @JsonInclude(JsonInclude.Include.NON_NULL)
    record NotificationRes(
        Long id,
        NotificationType type,
        String title,
        String description,
        LocalDateTime time,
        @JsonProperty("isRead") Boolean read,
        NotificationIcon iconName,
        NotificationColor iconColorHex,
        String email
    ){
        public static NotificationRes from(NotificationEntity notificationEntity) {
            return new NotificationRes(
                notificationEntity.getId(),
                notificationEntity.getType(),
                notificationEntity.getTitle(),
                notificationEntity.getDescription(),
                notificationEntity.getTime(),
                notificationEntity.isRead(),
                notificationEntity.getIconName(),
                notificationEntity.getIconColorHex(),
                notificationEntity.getEmail()
            );
        }

        public static NotificationRes fromWithoutEmail(NotificationEntity notificationEntity) {
            return new NotificationRes(
                notificationEntity.getId(),
                notificationEntity.getType(),
                notificationEntity.getTitle(),
                notificationEntity.getDescription(),
                notificationEntity.getTime(),
                notificationEntity.isRead(),
                notificationEntity.getIconName(),
                notificationEntity.getIconColorHex(),
                null // explicitly null for email
            );
        }

    }

}

File: src/main/kotlin/com/lcaohoanq/sp/domains/notifications/NotificationPort.kt
package com.lcaohoanq.sp.domains.notifications

import com.fasterxml.jackson.annotation.JsonInclude
import com.lcaohoanq.sp.entities.NotificationEntity
import com.lcaohoanq.sp.enums.NotificationEnum.*
import java.sql.Timestamp

interface NotificationPort {
    @JsonInclude(JsonInclude.Include.NON_NULL)
    data class NotificationRes(
        val id: Long?,
        val type: NotificationType?,
        val title: String?,
        val description: String?,
        val time: Timestamp?,
        val read: Boolean?,
        val iconName: NotificationIcon?,
        val iconColorHex: NotificationColor?,
        val email: String?
    ) {
        companion object {
            fun from(notificationEntity: NotificationEntity): NotificationRes {
                return NotificationRes(
                    notificationEntity.id,
                    notificationEntity.type,
                    notificationEntity.title,
                    notificationEntity.description,
                    notificationEntity.createdAt,
                    notificationEntity.read,
                    notificationEntity.iconName,
                    notificationEntity.iconColorHex,
                    notificationEntity.user?.email
                )
            }

            fun fromWithoutEmail(notificationEntity: NotificationEntity): NotificationRes {
                return NotificationRes(
                    notificationEntity.id,
                    notificationEntity.type,
                    notificationEntity.title,
                    notificationEntity.description,
                    notificationEntity.createdAt,
                    notificationEntity.read,
                    notificationEntity.iconName,
                    notificationEntity.iconColorHex,
                    null // explicitly null for email
                )
            }
        }
    }
}
